File: problems/LeetCode/60 days challenge/day10/p141.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        hare = head
        tortoise = head
        while tortoise and hare.next:
            hare = hare.next.next
            tortoise = tortoise.next
            if hare.val == tortoise.val:
                return True

        return False

File: LeetCode-Java/p141.java
class Solution {
    public class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }
    
    public boolean hasCycle(ListNode head) {
        if (head == null) return false;
        
        ListNode tortoise = head;
        ListNode hare = head;
        
        while (hare != null && hare.next != null) {
            hare = hare.next.next;
            tortoise = tortoise.next;
            
            if (hare == tortoise) {
                return true;
            }
        }
        
        return false;
    }
}
